model/map/src/main/java/org/keycloak/models/map/user/MapUserEntity.java

```diff
@@ -44,6 +44,8 @@ abstract class AbstractUserEntity extends Impl implements MapUserEntity {
 
         private static final Logger LOG = Logger.getLogger(MapUserProvider.class);
         private String id;
+        private String username;
+        private String usernameLowerCase;
```

Comment:
> If I understand correctly, is this field only used when serializing to JSON? 

No, actually the field is used only when querying by username (in case of `case-insensitive` search).

> Not sure how difficult is to compute lowercase version of String

Not sure exactly myself atm, it'd have to be updated in classes which actually parses the changesets to allow include a function as well (if it is not allowed already)

> but if we add a getter we could use it also in `MapFieldPredicates` as I suggested in the commit I sent you yesterday.

I'm not sure I follow here. Do you mean:

```
put(USER_PREDICATES, UserModel.SearchableFields.USERNAME, u -> (u instanceof MapUserEntityImpl) ? ((MapUserEntityImpl) u).getUsernameLowerCase() : KeycloakModelUtils.toLowerCaseSafe(u.getUsername()));
```
? If so I'd rather keep current state as it seems to me much cleaner. Maybe I am missing something here.

Reply:
> > If I understand correctly, is this field only used when serializing to JSON?
> 
> No, actually the field is used only when querying by username (in case of `case-insensitive` search).

Hmm, and when is the lowercase version of the username stored in the database? I thought the `usernameLowercase` field is somehow projected to `usernameLowercase` within the database and that is why we need it.

> > but if we add a getter we could use it also in `MapFieldPredicates` as I suggested in the commit I sent you yesterday.
> 
> I'm not sure I follow here. Do you mean:
> 
> ```
> put(USER_PREDICATES, UserModel.SearchableFields.USERNAME, u -> (u instanceof MapUserEntityImpl) ? ((MapUserEntityImpl) u).getUsernameLowerCase() : KeycloakModelUtils.toLowerCaseSafe(u.getUsername()));
> ```
> 
> ? If so I'd rather keep current state as it seems to me much cleaner. Maybe I am missing something here.

Well, in this case, the lowercase version of the username is computed each time the searchable field is used and not only in the CHM but also in HotRod as MapFieldPredicates are partially used also there. But I am not sure how big the problem is.

Reply:
> Hmm, and when is the lowercase version of the username stored in the database? I thought the `usernameLowercase` field is somehow projected to `usernameLowercase` within the database and that is why we need it.

Sorry, maybe I didn't express myself good enough. As the field is part of the `MapUserEntity` and the `JpaUserMetadata` extends `MapUserEntityImpl` it is already part of the json. Then it's generated into separate column in the database. And the column is then used in the search. Hope it makes more sense.

> Well, in this case, the lowercase version of the username is computed each time the searchable field is used and not only in the CHM but also in HotRod as MapFieldPredicates are partially used also there. But I am not sure how big the problem is.

If it'd be a problem I'd prefer to use the approach with `IgnoreMissingSetterForEntityClonerGenerator` annotation and use 
```
put(USER_PREDICATES, UserModel.SearchableFields.USERNAME, MapUserEntity::getUsernameLowerCase);
```